circus-train-s3-s3-copier/src/test/java/com/hotels/bdp/circustrain/s3s3copier/S3S3CopierTest.java

```diff
@@ -402,25 +404,97 @@ public void copySafelyShutDownTransferWhenRetryFails() throws Exception {
     TransferProgress transferProgress = new TransferProgress();
     when(copy.getProgress()).thenReturn(transferProgress);
     S3S3Copier s3s3Copier = new S3S3Copier(sourceBaseLocation, sourceSubLocations, replicaLocation, s3ClientFactory,
-        mockedRetryableTransferManagerFactory, listObjectsRequestFactory, registry, s3S3CopierOptions);
+        mockedTransferManagerFactory, listObjectsRequestFactory, registry, s3S3CopierOptions);
     try {
       s3s3Copier.copy();
       fail("Exception should have been thrown");
     } catch (CircusTrainException e) {
-      verify(mockedRetryableTransferManager).shutdownNow();
+      verify(mockedTransferManager).shutdownNow();
       assertThat(e.getMessage(), is("Error in S3S3Copier:"));
       assertThat(e.getCause().getMessage(), startsWith("S3 error"));
     }
   }
+
+  @Test
+  public void copyRetryOnlyFailedCopyJobs() throws InterruptedException {
+    String sourceKey1 = "bar/data1";
+    String sourceKey2 = "bar/data2";
+    client.putObject("source", sourceKey1, inputData);
+    client.putObject("source", sourceKey2, inputData);
+    Path sourceBaseLocation = new Path("s3://source/bar/");
+    Path replicaLocation = new Path("s3://target/foo/");
+    List<Path> sourceSubLocations = new ArrayList<>();
+
+    TransferManagerFactory mockedTransferManagerFactory = Mockito.mock(
+        TransferManagerFactory.class);
+    TransferManager mockedTransferManager = Mockito.mock(TransferManager.class);
+    when(mockedTransferManagerFactory.newInstance(any(AmazonS3.class), eq(s3S3CopierOptions)))
+        .thenReturn(mockedTransferManager);
+    Copy copy = Mockito.mock(Copy.class);
+    when(mockedTransferManager.copy(any(CopyObjectRequest.class), any(AmazonS3.class),
+        any(TransferStateChangeListener.class))).thenReturn(copy);
+    when(copy.getProgress()).thenReturn(new TransferProgress());
+    doThrow(new AmazonClientException("cause"))
+        .doNothing().when(copy).waitForCompletion();
+    S3S3Copier s3s3Copier = new S3S3Copier(sourceBaseLocation, sourceSubLocations, replicaLocation, s3ClientFactory,
+        mockedTransferManagerFactory, listObjectsRequestFactory, registry, s3S3CopierOptions);
+    try {
+      s3s3Copier.copy();
+      ArgumentCaptor<CopyObjectRequest> captor = ArgumentCaptor.forClass(CopyObjectRequest.class);
+      verify(mockedTransferManager, Mockito.times(3)).copy(captor.capture(), any(AmazonS3.class), any(TransferStateChangeListener.class));
+      List<CopyObjectRequest> capturedCopyRequests = captor.getAllValues();
+      assertThat(capturedCopyRequests.get(0).getSourceKey(), is(sourceKey1));
+      assertThat(capturedCopyRequests.get(1).getSourceKey(), is(sourceKey2));
+      assertThat(capturedCopyRequests.get(2).getSourceKey(), is(sourceKey1));
+      verify(mockedTransferManager).shutdownNow();
+      verifyNoMoreInteractions(mockedTransferManager);
+    } catch (CircusTrainException e) {
+      fail("Exception should not have been thrown");
```

Comment:
@patduin feel free to chip in here, but we could remove the try catch and let this fail the test if thrown without having `fail` - at least that way we would see the stack trace of `e` which would give us a better idea of why it failed.

Reply:
I think the intent of this test is to show that if an exception is thrown the exception is caught and the copy is retried and then everything is ok. If the exception is bubbled up the retry logic has failed, which is what the tests nicely documents like this.
There are other tests where the copy succeeds and no retry logic is used, those will throw the exception you are after and appropriate stacktrace